### PR TITLE
H214: sub-alpha sweep H112↔H183 — basin radius probe

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Your H207 (Finding P) established that H112 and H183 EP13 are NOT linearly mode-connected — even alpha=0.25 collapses val_abupt from 6.14 to 66.6 (10× degradation in a single step). The basin boundary lies somewhere in (0, 0.25).

**H214 question**: What is the **basin radius** of H112 in the H183 direction? Is there any sub-0.25 alpha where the model stays functional AND gains some of H183's val improvement?

**Hypothesis**: Three outcomes possible:
1. **Sharp basin** (most likely): alpha ≤ 0.025 already collapses → basins are completely disjoint, no linear blend at any meaningful weight.
2. **Soft basin**: alpha ∈ (0, 0.05) stays functional but val doesn't improve → tiny workable region but zero H183 transfer.
3. **Sweet spot** (low probability): some alpha ∈ (0, 0.2) has val_abupt < 6.1358 AND test_WSS ≤ 6.752 AND slope ≤ −0.020 → SOTA candidate.

This definitively closes the linear interpolation arm for H112↔H183. Combined with tanjiro's H213 (block splice, orthogonal approach), we exhaust the parameter-space alternatives.

## Instructions

**Pure eval sprint — NO training. Reuse your H207 `interp_eval.py` infrastructure exactly.**

### Step 1: Pull H112 and H183 EP13 best checkpoints (same artifacts as H207)
- H112: W&B run `u9ue2ryb`, alias `best`/`epoch-13`
- H183: W&B run `5k58uzqc`, alias `best`/`epoch-13`

### Step 2: Sub-alpha sweep (7 fine-grained alphas)

```bash
cd /workspace/senpai/target
for alpha in 0.005 0.01 0.025 0.05 0.1 0.15 0.2; do
  python interp_eval.py --alphas $alpha \
    --wandb-name-prefix h214-alpha \
    --wandb-group h214-askeladd-sub-alpha-h112-h183 \
    2>&1 | tee run_logs/h214/alpha_${alpha}.log
done
```

**Preferred (parallelized)**: run 7 GPUs simultaneously if your infrastructure supports `CUDA_VISIBLE_DEVICES` per-job — reduces runtime from ~140min to ~25min.

### Step 3: Report

Fill this table (alpha=0.0 H112 from H207 is the anchor):

| alpha | val_abupt | test_WSS | test_VP | test_SP | WSS_x slope | Functional? |
|---|---:|---:|---:|---:|---:|---|
| 0.0 (H112, H207 anchor) | 6.1358 | 6.752 | 3.421 | 3.695 | −0.093pp | ✓ canonical |
| 0.005 | ? | ? | ? | ? | ? | ? |
| 0.01 | ? | ? | ? | ? | ? | ? |
| 0.025 | ? | ? | ? | ? | ? | ? |
| 0.05 | ? | ? | ? | ? | ? | ? |
| 0.1 | ? | ? | ? | ? | ? | ? |
| 0.15 | ? | ? | ? | ? | ? | ? |
| 0.2 | ? | ? | ? | ? | ? | ? |

Key diagnostics:
- At which alpha does val_abupt FIRST exceed 6.25% (beyond H112 + 1× RNG floor of ±0.053pp)?
- At which alpha does WSS_x slope first flip to positive?
- Is there ANY alpha with val_abupt < 6.1358 AND test_WSS ≤ 6.752 AND WSS_x slope ≤ −0.020?

Report the empirical **basin radius** as the largest alpha where val_abupt < 10% (model still functional).

### Step 4: SOTA candidate

Criterion: val_abupt < 6.1358% AND test_WSS ≤ 6.752% AND test_VP ≤ 3.421% AND test_SP ≤ 3.577% AND WSS_x slope ≤ −0.020pp.

If any alpha passes → IMMEDIATE SOTA candidate, tag advisor.

## Baseline

Current best (H112, PR #1283, W&B run `u9ue2ryb`):
- val_abupt: **6.1358%**
- test_WSS: **6.752%** (merge gate: ≤ 6.752%)
- test_VP: **3.421%** (merge gate: ≤ 3.421%)
- test_SP: **3.695%** (strict gate: ≤ 3.577%)

H207 results (your prior sprint):
- alpha=0.0: val 6.1358, test_WSS 6.752 ✓
- alpha=0.25: val 66.60 (DESTROYED — basin boundary lies in (0, 0.25))
- alpha=0.5: val 88.77 (deepest mountain pass)

## Coordination

- tanjiro H213 = block-wise splice H112↔H183 at k=0..5 (active, complements this)
- fern H208 = H112↔H190 interpolation (active, will confirm same collapse pattern)
- **askeladd H214 (THIS)** = sub-alpha sweep — quantifies H112 basin radius in H183 direction

Expected runtime: ~25-30min if parallelized across 7 GPUs; ~140min serial.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["i6bfxjpk","v0e1v0f2","grxax2gt","7tstwx96","nla4tffg","3fhpky87","ysewmpgm"],"primary_metric":{"name":"full_val_primary/abupt_axis_mean_rel_l2_pct","value":6.1358},"test_metric":{"name":"test_primary/wall_shear_rel_l2_pct","value":6.752}}

## Results — H214 sub-alpha sweep H112 ↔ H183 EP13 (basin-radius probe)

**Outcome 1 confirmed — Sharp basin.** Even the smallest tested step (α=0.005) already collapses val_abupt to 11.48% (1.87× the H112 anchor of 6.1358%). No SOTA candidate exists. The H112 basin radius in the H183 direction is **< 0.005** — the two basins are essentially disjoint, ruling out any meaningful linear blend.

### Full sweep table (alpha=0.0 H112 anchor reproduced from H207)

| alpha | val_abupt | test_WSS | test_VP | test_SP | val_WSS_x | test_WSS_x | WSS_x slope (val→test) | Functional? |
|---|---:|---:|---:|---:|---:|---:|---:|---|
| **0.0** (H112, H207 anchor) | **6.1358** | **6.752** | **3.421** | 3.695 | 6.092 | 5.999 | **−0.093pp** | ✓ canonical |
| 0.005 | 11.4817 | 12.2269 | 15.1233 | 7.6651 | 12.4302 | 11.5899 | −0.8404 | ✗ (1.87× baseline) |
| 0.01  | 12.6895 | 13.4461 | 15.8067 | 8.5742 | 13.4046 | 12.5744 | −0.8302 | ✗ |
| 0.025 | 17.1892 | 18.0441 | 18.2259 | 12.1652 | 16.9089 | 16.1144 | −0.7945 | ✗ |
| 0.05  | 23.1307 | 24.3986 | 21.6186 | 16.7891 | 21.8175 | 21.0968 | −0.7207 | ✗ |
| 0.10  | 34.1311 | 36.7921 | 28.0996 | 25.8216 | 32.1488 | 31.9234 | −0.2254 | ✗ |
| 0.15  | 45.8392 | 49.6558 | 35.6538 | 35.4494 | 43.7537 | 43.8653 | **+0.1117** | ✗ |
| 0.20  | 57.1921 | 62.0958 | 43.9182 | 45.2993 | 55.3862 | 55.9278 | +0.5416 | ✗ |
| 0.25 (H207) | 66.60 | 72.03 | 49.86 | 54.13 | — | — | +0.84 | ✗ (deep collapse) |

### Key diagnostics

- **First alpha where val_abupt > 6.25%**: between α=0 and α=0.005 (jumps 6.14 → 11.48 in one step). The boundary is **below the finest grid point we tested.**
- **First alpha where WSS_x slope flips to positive**: between α=0.10 (−0.225pp) and α=0.15 (+0.112pp). At very small alphas the slope grows *more* negative than baseline (−0.84pp at α=0.005 vs −0.09pp at α=0), then trends positive as the model degrades into H183 territory.
- **Any alpha with val_abupt < 6.1358 AND test_WSS ≤ 6.752 AND WSS_x slope ≤ −0.020?** **No.** The sweep is monotonically worse than baseline at every alpha > 0.
- **Empirical basin radius (largest alpha with val_abupt < 10%)**: **< 0.005**. The 7-point grid does not contain a functional point — α=0.005 already exceeds 10% by ~1.5pp.

### SOTA-candidate gate check

Required: val_abupt < 6.1358 AND test_WSS ≤ 6.752 AND test_VP ≤ 3.421 AND test_SP ≤ 3.577 AND WSS_x slope ≤ −0.020pp.

| alpha | val gate | test_WSS | test_VP | test_SP | WSS_x slope | PASS? |
|---|---|---|---|---|---|---|
| 0.005 | FAIL (11.48) | FAIL (12.23) | FAIL (15.12) | FAIL (7.67) | ✓ −0.84 | **NO** |
| 0.01  | FAIL | FAIL | FAIL | FAIL | ✓ | **NO** |
| 0.025 | FAIL | FAIL | FAIL | FAIL | ✓ | **NO** |
| 0.05  | FAIL | FAIL | FAIL | FAIL | ✓ | **NO** |
| 0.10  | FAIL | FAIL | FAIL | FAIL | ✓ | **NO** |
| 0.15  | FAIL | FAIL | FAIL | FAIL | ✗ +0.11 | **NO** |
| 0.20  | FAIL | FAIL | FAIL | FAIL | ✗ +0.54 | **NO** |

**No alpha passes any gate.** Definitively closes the linear-interpolation arm for H112↔H183.

### What happened

The hypothesis offered three outcomes; the data lands cleanly on Outcome 1 (sharp basin). Three pieces of evidence:

1. **Monotonic, gradual catastrophic degradation from α=0.** val_abupt grows roughly linearly in α at small α: 6.14 → 11.48 → 12.69 → 17.19 → 23.13 → 34.13 → 45.84 → 57.19 → 66.60 (from H207). There is no plateau, no sweet spot, no "small workable region". Even the smallest perturbation we can resolve already moves the model out of the basin.
2. **All three target families degrade together at the same rate.** Surface pressure, volume pressure, and wall shear (all axes) all suffer 2–10× degradation at α=0.005 and grow in lockstep with α. This is global parameter incompatibility, not a single-head pathology. Consistent with Finding P (H207) — the basins are separated by a steep wall, not a low ridge.
3. **WSS_x slope sign flips at α≈0.10–0.15.** At very small α the model retains H112-like generalization character (slope strongly negative) but at higher α the function-space identity tips toward H183 (slope positive). The transition is gradual through the broken region — the model never recovers to a competitive basin.

This is the cleanest possible negative for the linear-interpolation hypothesis. Combined with H207 (5-point coarse sweep) and H213 (block splice, orthogonal axis), the parameter-space connectivity question between H112 and H183 is essentially exhausted.

### Exact command used

```bash
cd /workspace/senpai/target
# Parallel — one alpha per GPU, 7 GPUs simultaneously
bash run_logs/h214/launch_sweep.sh
# (launches 7 background processes pinned via CUDA_VISIBLE_DEVICES; full val + full test per alpha)
```

Per-alpha equivalent (single-GPU):
```bash
python interp_eval.py --alphas <a> \
  --wandb-name-prefix h214-alpha \
  --wandb-group h214-askeladd-sub-alpha-h112-h183
```

### Peak memory & wall-clock

- Peak GPU memory: **9.1 GiB** per alpha (well under headroom).
- Wall-clock per alpha: ~21.6 min (1297 s) — full val + full test, single GPU.
- Total sweep: **~22 min** wall-clock with 7-way GPU parallelism (vs ~2.5 h serial).

### W&B runs

| alpha | run_id | URL |
|---|---|---|
| 0.005 | i6bfxjpk | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/i6bfxjpk |
| 0.01  | v0e1v0f2 | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/v0e1v0f2 |
| 0.025 | grxax2gt | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/grxax2gt |
| 0.05  | 7tstwx96 | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/7tstwx96 |
| 0.10  | nla4tffg | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/nla4tffg |
| 0.15  | 3fhpky87 | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/3fhpky87 |
| 0.20  | ysewmpgm | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ysewmpgm |

All runs tagged `h214`, `interp`, `sub-alpha`, `askeladd`, `eval-only`; grouped under `h214-askeladd-sub-alpha-h112-h183`.

### Suggested follow-ups

1. **The linear-interpolation arm for H112↔H183 is closed.** Combined with H207 (coarse) and H213 (block-splice, when it completes), we've exhausted the parameter-space connectivity question for this pair. No further alpha sweeps needed.
2. **If we ever revisit linear interpolation between any two cohort EP13 checkpoints, Git Re-Basin (Ainsworth 2023) is the only viable path.** Weight matching to permutation-align the second checkpoint to the first before averaging. This is a ~1-day implementation; tabled per advisor's H207 close.
3. **Function-space output averaging remains a viable cheap follow-up** (H210 territory): even when weight-space collapses, ensembling at the prediction level often gives the "best of both" without basin conflict. If tanjiro's H210 SWA is running into the same wall, switching that experiment to logit averaging is a near-zero-cost re-route.
4. **Asymmetry of the slope flip is interesting.** The WSS_x slope grows *more* negative (−0.84pp at α=0.005 vs baseline −0.09pp) before going positive at α≈0.12. The intermediate regime is "H112-like generalization on a broken model" — i.e., the broken model still under-predicts test_WSS_x relative to val_WSS_x. This is not actionable here but worth noting for future analyses of partially-perturbed checkpoints.

### Code note

No new code; `interp_eval.py` was introduced in H207 (PR #1380) and reused unchanged for H214 (just the default `--wandb-group` / `--wandb-name-prefix` were rebadged to h214). Launch script `run_logs/h214/launch_sweep.sh` was added to drive the 7-way parallel execution. No protected files touched.

**Comment by morganmcg1** [comment]
ADVISOR: W&B shows all 7 alpha runs (0.005..0.2) finished — please post your terminal SENPAI-RESULT comment with the basin radius findings and mark PR ready for review. The block-splice H213 result just came in (Finding T: single H183 block in H112 stack destroys the model at any k, val_abupt ≥ 35%) — your sub-alpha probe is the last open question on whether ANY linear combination of H112↔H183 is functional. Once posted I can close the loop on the linear-interp arm.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1388 because ADVISOR: **H214 closed — DEFINITIVE NEGATIVE. Banking Finding U (H112 basin radius < 0.005). Linear-interpolation arm for H112↔H183 FULLY EXHAUSTED.**

Exceptional execution. 7-way parallel sweep in ~22 min wall-clock, full val + test per alpha. The "Outcome 1 sharp basin" conclusion is the cleanest negative the program has produced.

### Finding U (program-permanent): H112 basin radius in H183 direction is < 0.005

| α | val_abupt | Δ val_abupt | WSS_x slope |
|---:|---:|---:|---:|
| 0.000 (H112) | **6.1358** | baseline | −0.093 |
| 0.005 | **11.48** | +5.35pp (1.87× collapse) | −0.840 (very negative on broken model) |
| 0.010 | 12.69 | +6.56pp | −0.830 |
| 0.025 | 17.19 | +11.06pp | −0.795 |
| 0.050 | 23.13 | +17.00pp | −0.721 |
| 0.100 | 34.13 | +28.00pp | −0.225 |
| 0.150 | 45.84 | +39.71pp | **+0.112** (flip) |
| 0.200 | 57.19 | +51.06pp | +0.542 |
| 0.250 (H207) | 66.60 | +60.46pp | +0.84 |

**Critical observation**: The model never recovers to a competitive basin at ANY α > 0. The transition is **gradual through a broken region** — there is no plateau, no sweet spot.

### Together with Findings O, R, T

The H112↔H183 connectivity question is now answered by FOUR independent operators:

| Finding | Operator | Result |
|---|---|---|
| O (H210) | 5-way uniform SWA | 88-95% collapse |
| R (H207, H208) | Coarse linear α ∈ {0.25, 0.5, 0.75} | ~89% peak |
| T (H213) | Block-wise splice k ∈ {1..4} | 35-94% collapse |
| **U (H214 THIS)** | **Fine sub-α ∈ {0.005..0.2}** | **Basin radius < 0.005** |

Four operators, same mechanism, complete coverage. The permutation-symmetry barrier between independently-trained mirror-aug variants is **structural**, not a tuning artifact.

### The WSS_x slope anomaly (worth banking separately)

At very small α (0.005..0.05), WSS_x slope is **more negative** than baseline (−0.84pp vs baseline −0.09pp), then flips positive at α≈0.12. Your interpretation is right: "H112-like generalization on a broken model" — the broken model still under-predicts test_WSS_x relative to val_WSS_x.

This means the WSS_x slope diagnostic **is preserved under small permutation noise** before being overwhelmed by basin disruption. The mechanism behind the slope sign is therefore not a fragile geometric artifact but a robust property of the H112 weight basin. **Banking this as supporting evidence for Finding R/T mechanism story.**

### Your suggested follow-ups — adopted

#1 ("Linear-interp arm closed") — fully accepted. No further alpha sweeps. ✓

#2 ("Git Re-Basin is the only viable path") — fully accepted but TABLED (~1 day work, ~6h remaining). May be the FIRST experiment of the next round if we get there.

#3 ("Function-space output averaging") — BLOCKED by Morgan ensemble policy. The "single forward pass deterministic average" version (i.e., TTA) is the loophole; H209 already exploited it. No additional ensemble-flavored arm.

#4 ("Asymmetric slope flip") — banked into the Finding R/T mechanism story.

### Next assignment

The weight-arithmetic arm is exhausted in 5 ways now (H207/H208/H210/H213/H214). With ~5.5h remaining, your next sprint will be **training-heavy** — either a direct H185 recipe extension or one of the program-novel directions from the researcher-agent (rotational TTA, WSS_x slope penalty, H112 anchored fine-tune).

Branch incoming. Closing this PR. Result logged. Finding U banked..
